pkg/gofr/datasource/opentsdb/response.go

```diff
@@ -0,0 +1,416 @@
+}
+
+func (queryLastResp *QueryLastResponse) getCustomParser(ctx context.Context, logger Logger,
+	tracer trace.Tracer) func(respCnt []byte) error {
+	return queryParserHelper(ctx, logger, tracer, queryLastResp, "GetCustomParser-QueryLast")
+}
+
+func (verResp *VersionResponse) getCustomParser(ctx context.Context, logger Logger, tracer trace.Tracer) func(respCnt []byte) error {
+	return customParserHelper(ctx, verResp, "GetCustomParser-VersionResp", logger, tracer,
+		func(resp []byte) error {
+			v := make(map[string]any, 0)
+
+			err := json.Unmarshal(resp, &v)
+			if err != nil {
+				return err
+			}
+
+			verResp.VersionInfo = v
+
+			return nil
+		})
+}
+
+func (annotResp *AnnotationResponse) getCustomParser(ctx context.Context, logger Logger, tracer trace.Tracer) func(respCnt []byte) error {
+	return customParserHelper(ctx, annotResp, "getCustomParser-Annotation", logger, tracer,
+		func(resp []byte) error {
+			if len(resp) == 0 {
+				return nil
+			}
+
+			return json.Unmarshal(resp, &annotResp)
+		})
+}
+
+func (aggreResp *AggregatorsResponse) getCustomParser(ctx context.Context, logger Logger, tracer trace.Tracer) func(respCnt []byte) error {
+	return customParserHelper(ctx, aggreResp, "GetCustomParser-Aggregator", logger, tracer,
+		func(resp []byte) error {
+			j := make([]string, 0)
+
+			err := json.Unmarshal(resp, &j)
+			if err != nil {
+				return err
+			}
+
+			aggreResp.Aggregators = j
+
+			return nil
+		})
+}
+
+type genericResponse interface {
+	addTrace(ctx context.Context, tracer trace.Tracer, operation string) trace.Span
+}
+
+// sendRequest dispatches an HTTP request to the OpenTSDB server, using the provided
+// method, URL, and body content. It returns the parsed response or an error, if any.
+func (c *Client) sendRequest(ctx context.Context, method, url, reqBodyCnt string, parsedResp response) error {
+	span := c.addTrace(ctx, "sendRequest")
+
+	status := statusFailed
+
+	var message string
+
+	defer sendOperationStats(c.logger, time.Now(), "sendRequest", &status, &message, span)
+
+	// Create the HTTP request, attaching the context if available.
+	req, err := http.NewRequest(method, url, strings.NewReader(reqBodyCnt))
+	if ctx != nil {
+		req = req.WithContext(ctx)
+	}
+
+	if err != nil {
+		errRequestCreation := fmt.Errorf("failed to create request for %s %s: %w", method, url, err)
+
+		message = fmt.Sprint(errRequestCreation)
+
+		return errRequestCreation
+	}
+
+	// Set the request headers.
+	req.Header.Set("Content-Type", "application/json; charset=UTF-8")
+
+	// Send the request and handle the response.
+	resp, err := c.client.Do(req)
+	if err != nil {
+		errSendingRequest := fmt.Errorf("failed to send request for %s %s: %w", method, url, err)
+
+		message = fmt.Sprint(errSendingRequest)
+
+		return errSendingRequest
+	}
+
+	if resp.StatusCode >= 400 && resp.StatusCode < 500 {
+		return fmt.Errorf("client error: %d", resp.StatusCode)
+	}
+
+	defer resp.Body.Close()
```

Comment:
The body might need to be closed even on 400 and 500

This could be enough

```suggestion
	defer resp.Body.Close()
	if resp.StatusCode >= 400 && resp.StatusCode < 500 {
		return fmt.Errorf("client error: %d", resp.StatusCode)
	}
```
But this would be safer, I think

```suggestion
	if resp.Body != nil {
		defer resp.Body.Close()
	}

	if resp.StatusCode >= 400 && resp.StatusCode < 500 {
		return fmt.Errorf("client error: %d", resp.StatusCode)
	}
```